// image-upscaler/image-upscaler.interface.ts
export type ImageFormat = "png" | "jpg" | "webp" | "";

export type ModelName =
  | "realesrgan-x4fast"
  | "realesrgan-x4plus-anime"
  | "realesrgan-x4plus"
  | "remacri"
  | "ultramix_balanced"
  | "ultrasharp";

export interface File {
  fieldname: string;
  originalname: string;
  encoding: string;
  mimetype: string;
  destination: string;
  filename: string;
  path: string;
  size: number;
}

export interface Options {
  inputDir: string;
  outputDir: string;
  modelsPath: string;
  model: string;
  gpuId: string;
  saveImageAs: ImageFormat;
  scale: string;
  customWidth: string;
  compression: string;
  tileSize: string;
  dimension: string;
}

export interface DimensionOptions {
  image: Buffer;
  saveImageAs: ImageFormat;
  model: ModelName;
  width: number;
  height: number;
  timeout?: number;
}

export interface ScaleOptions {
  image: Buffer;
  saveImageAs: ImageFormat;
  model: ModelName;
  scale: string;
  timeout?: number;
}
export interface WidthOptions {
  image: Buffer;
  saveImageAs: ImageFormat;
  model: ModelName;
  width: string;
  timeout?: number;
}
